### ISSUE TITLE
Do/should we have a convention about reassigning to argument aliases vs. making new ones?

### ISSUE THREAD

**Issue by gibson042** [tc39/ecma262#2967]
(prompted by https://github.com/tc39/proposal-temporal/pull/2438#discussion_r1037144147 )

## Primary issue

_Many_ operations have a need to coerce their arguments via ℝ/​RequireObjectCoercible/​StringToCodePoints/​ToIntegerOrInfinity/​ToNumber/​ToString/​etc., or to replace sentinel values with defaults. Some of them do so by reassigning to parameter aliases, like [BigIntBitwiseOp](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-bigintbitwiseop) and [SetRealmGlobalObject](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-setrealmglobalobject) and [OrdinarySetWithOwnDescriptor](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-ordinarysetwithowndescriptor). Others achieve the same goal by introducing new aliases, such as Date.prototype methods like [setMinutes](https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-date.prototype.setminutes) (with notable exceptions [setMilliseconds](https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-date.prototype.setmilliseconds) and [setUTCMilliseconds](https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-date.prototype.setutcmilliseconds), which use reassignment), String.prototype methods like [codePointAt](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.codepointat), RegExp.prototype methods like [exec](https://tc39.es/ecma262/multipage/text-processing.html#sec-regexp.prototype.exec), Array.prototype methods like [join](https://tc39.es/ecma262/multipage/indexed-collections.html#sec-array.prototype.join), and so on.

Some operations even mix styles:
* [ApplyStringOrNumericBinaryOperator](https://tc39.es/ecma262/multipage/ecmascript-language-expressions.html#sec-applystringornumericbinaryoperator) coerces arguments _lval_ and _rval_ into aliases _lprim_ and _rprim_, but then overwrites the former pair with the latter.
* String.prototype methods [replace](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replace) and [replaceAll](https://tc39.es/ecma262/multipage/text-processing.html#sec-string.prototype.replaceall) and [RegExp.prototype[@@replace]](https://tc39.es/ecma262/multipage/text-processing.html#sec-regexp.prototype-@@replace) coerce argument _searchValue_ into alias _searchString_, but reassign to argument _replaceValue_.
* [GetViewValue](https://tc39.es/ecma262/multipage/structured-data.html#sec-getviewvalue) and [SetViewValue](https://tc39.es/ecma262/multipage/structured-data.html#sec-setviewvalue) coerce argument _requestIndex_ into alias _getIndex_, but reassign argument _isLittleEndian_.
* [JSON.stringify](https://tc39.es/ecma262/multipage/structured-data.html#sec-json.stringify) reassigns to argument _space_ before using it to derive alias _gap_.
* [Reflect.get](https://tc39.es/ecma262/multipage/reflection.html#sec-reflect.get) and [Reflect.set](https://tc39.es/ecma262/multipage/reflection.html#sec-reflect.set) coerce argument _propertyKey_ to alias _key_ but default optional argument _receiver_ by assignment.

## Proposed solution

I don't think this is practical to lint, but I am of the opinion that _generally_ it is better to reassign to the parameter rather than inventing two names for the same concept in an algorithm—naming _once_ is hard enough!

## Further observations

Some operations reassign parameter aliases later in their algorithm:
* [CopyDataBlockBytes](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-copydatablockbytes) uses arguments _toIndex_, _fromIndex_, and _count_ as iteration variables.
* [Object.prototype.isPrototypeOf](https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-object.prototype.isprototypeof) reassigns argument _V_ as it walks the prototype chain.
* [SetFunctionName](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-setfunctionname) imperatively updates _name_ based on both its type and the presence of a nonempty _prefix_ argument.

I think these should be handled case by case, but I have a moderate preference for introducing new aliases rather than arbitrarily modifying arguments.

**Comment by ljharb** [comment]
I share the opinion that i'd prefer a new alias/variable name than reassigning to an existing one.

**Comment by ptomato** [comment]
This seems like an ideal kind of guidance to put in a document like https://github.com/tc39/how-we-work/pull/119, if that ever moves forward.

**Comment by jmdyck** [comment]
This issue overlaps with issue #2178. In particular, see https://github.com/tc39/ecma262/issues/2178#issuecomment-701676969, which conveys the editorial consensus that "We should never assign to parameters".

**Comment by gibson042** [comment]
Thanks very much @jmdyck. If the editors are still of that opinion, then it seems to me like sufficiently strong motivation to establish alias naming conventions to be used in coercion and derivation.

**Comment by gibson042** [comment]
Some possibilities (not necessarily exclusive):
* Capitalize parameter names, e.g. `Let _x_ be ℝ(_X_)` and `Let _position_ be ? ToIntegerOrInfinity(_Position_)`.
* Prefix alias names, e.g. `Let _bLittleEndian_ be ToBoolean(_littleEndian_)`.
* Suffix alias names, e.g. `Let _searchStr_ be ? ToString(_searchString_)` and `Let _posEMV_ be ? ToIntegerOrInfinity(_pos_)`.

**Comment by ljharb** [comment]
I’d rather try to pick names that are more descriptive, if we can

**Comment by jmdyck** [comment]
@ljharb More descriptive how? In the cases considered here, we've got 2 aliases where the intent/purpose is the same, it's just that the second results from coercing the first somehow. So if you want to distinguish the aliases by picking more descriptive names, it seems like the only thing you can be more descriptive *about* is the type of the value. Are you suggesting that (in these cases) the aliases should more explicitly describe (the types of) the values that they can refer to? E.g.:
- `_positionAsAnECMAScriptLanguageValue_` vs `_positionAsAnIntegerOrInfinity_` (maybe `_positionValue_` vs `_positionIntOrInf_`)
- `_littleEndianValue_` vs `_littleEndianBoolean_`
- `_searchValue_` vs `_searchString_`

?

**Comment by ljharb** [comment]
I don’t have a silver bullet answer in mind - but there’s lots of options like, eg, isLittleEndian, that don’t have to follow a notational rule but can just make sense for the usage.

**Comment by gibson042** [comment]
> I’d rather try to pick names that are more descriptive, if we can

@ljharb Does that mean that you support prefixing and/or suffixing as strategies but just aren't happy with the [Hungarian](https://en.wikipedia.org/wiki/Hungarian_notation)-like examples?

> there’s lots of options like, eg, isLittleEndian, that don’t have to follow a notational rule but can just make sense for the usage.

Assuming _is\<Adjective>_ would be used rather than e.g. _b\<Adjective>_ for aliases with value limited to the Boolean language type, that doesn't seem to be meaningfully different from "a notational rule" (even if also allowing for _are\<Adjective>_/_was\<Adjective>_/etc.). And more importantly, what pattern would be used for the harder cases of mathematical and string coercion? As @jmdyck points out, these are cases where the purpose of multiple aliases is identical and the only distinction between them is pre- vs. post-coercion to a target data type, and the current text makes clear that if reassignment is discouraged then the absence of prescribed patterns results in a plethora of unhelpful names.

**Comment by ljharb** [comment]
I specifically didn't say "prefix booleans with is", i gave one ad hoc example :-)

I'm saying that there actively should be *no rule*, and that each algorithm should pick whatever names subjectively feel like the best fit.

**Comment by jmdyck** [comment]
@ljharb I think maybe you're answering a different question than the one that @gibson042 and I are discussing.
E.g., when I say
- `_littleEndianValue_` vs `_littleEndianBoolean_`

I **don't** mean "We have to choose a name for this alias, here are two possibilities" (where `_isLittleEndian_` would be a perfectly acceptable alternative).

I mean "We have to choose names for 2 closely related aliases, how shall we distinguish them?" (So suggesting `_isLittleEndian_` just raises the questions "For which one (pre or post coercion)? And what would the other be called?")

**Comment by ljharb** [comment]
Right, I get that - I don't think a general rule is the right way to approach that tho, and that each choice should be ad hoc.
